### PR TITLE
Fixes #5532 null pointer dereference in EmitBooleanExpression 

### LINKED ISSUE #5532: null pointer dereference in EmitBooleanExpression
the following js code fragment can lead to a null pointer dereference.
```
function g()
{
    for (var i=0;;++i)
    {
        for (var j=0;;++j){
             var a =0;
        }
    }
} 

g();
```

==10216==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x55b703964384 bp 0x7ffd0824e480 sp 0x7ffd0824e430 T0)
    #0 0x55b703964383 in EmitBooleanExpression(ParseNode*, int, int, ByteCodeGenerator*, FuncInfo*, bool, bool) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x170a383)
    #1 0x55b703966025 in ByteCodeGenerator::EmitInvertedLoop(ParseNodeLoop*, ParseNodeFor*, FuncInfo*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x170c025)
    #2 0x55b70392a05b in Emit(ParseNode*, ByteCodeGenerator*, FuncInfo*, int, bool, ParseNode*, bool) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x16d005b)
    #3 0x55b703924439 in ByteCodeGenerator::EmitTopLevelStatement(ParseNode*, FuncInfo*, int) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x16ca439)
    #4 0x55b703942582 in ByteCodeGenerator::EmitFunctionBody(FuncInfo*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x16e8582)
    #5 0x55b703947d95 in ByteCodeGenerator::EmitOneFunction(ParseNodeFnc*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x16edd95)
    #6 0x55b703942fa4 in ByteCodeGenerator::EmitScopeList(ParseNode*, ParseNode*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x16e8fa4)
    #7 0x55b703942b77 in ByteCodeGenerator::EmitScopeList(ParseNode*, ParseNode*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x16e8b77)
    #8 0x55b703942f91 in ByteCodeGenerator::EmitScopeList(ParseNode*, ParseNode*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x16e8f91)
    #9 0x55b7029737a1 in ByteCodeGenerator::Generate(ParseNodeProg*, unsigned int, ByteCodeGenerator*, Js::ParseableFunctionInfo**, unsigned int, bool, Parser*, Js::ScriptFunction**) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x7197a1)
    #10 0x55b70297a6c8 in GenerateByteCode(ParseNodeProg*, unsigned int, Js::ScriptContext*, Js::ParseableFunctionInfo**, unsigned int, bool, Parser*, CompileScriptException*, Js::ScopeInfo*, Js::ScriptFunction**) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x7206c8)
    #11 0x55b7028eb29a in Js::ScriptContext::GenerateRootFunction(ParseNodeProg*, unsigned int, Parser*, unsigned int, CompileScriptException*, char16_t const*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x69129a)
    #12 0x55b7028eb63d in Js::ScriptContext::LoadScript(unsigned char const*, unsigned long, SRCINFO const*, CompileScriptException*, Js::Utf8SourceInfo**, char16_t const*, LoadScriptFlag, void*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x69163d)
    #13 0x55b7026c1fdf in RunScriptCore(void*, unsigned char const*, unsigned long, LoadScriptFlag, unsigned long, char16_t const*, bool, _JsParseScriptAttributes, bool, void**) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x467fdf)
    #14 0x55b7026cb85d in JsRun (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x47185d)
    #15 0x55b7025d1f97 in RunScript(char const*, char const*, unsigned long, void (*)(void*), void*, char*, void*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x377f97)
    #16 0x55b7025d3e72 in ExecuteTest(char const*) (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x379e72)
    #17 0x55b7025d49f2 in main (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x37a9f2)
    #18 0x7f70646fe82f in __libc_start_main /build/glibc-Cl5G7W/glibc-2.23/csu/../csu/libc-start.c:291
    #19 0x55b7024fbce8 in _start (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x2a1ce8)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/media/b/ext/ChakraCore-1.10.0/out/Release/ch+0x170a383) in EmitBooleanExpression(ParseNode*, int, int, ByteCodeGenerator*, FuncInfo*, bool, bool)
==10216==ABORTING

### PR DESCRIPTION
**Pull request by wyrichte**
Fixes #5532 
Given a nested for loop that qualifies for loop inversion, the lack of a conditional in the outer loop causes a read access violation when the conditional’s ParseNode’s nop is read. The conditional is analyzed to perform a zero trip test which can skip the execution of the loop entirely. In the case where the outer loop lacks a conditional, the zero trip test will not pass. Therefore when the outer loop lacks a conditional skipping the inclusion of a zero trip test is valid and avoids the read access violation.

### CODE REVIEW

**Comment by wyrichte** [comment]
@curtisman @pleath @akroshg @aneeshdk 
please review

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp [hunk 1]
```diff
@@ -9259,12 +9259,16 @@ void ByteCodeGenerator::EmitInvertedLoop(ParseNodeLoop* outerLoop, ParseNodeFor*
...
     this->m_writer.MarkLabel(invertedLoopLabel);
 
-    // Emit a zero trip test for the original outer-loop
-    Js::ByteCodeLabel zeroTrip = this->m_writer.DefineLabel();
-    ParseNode* testNode = this->GetParser()->CopyPnode(outerLoop->AsParseNodeFor()->pnodeCond);
-    EmitBooleanExpression(testNode, zeroTrip, afterInvertedLoop, this, funcInfo, true, false);
-    this->m_writer.MarkLabel(zeroTrip);
-    funcInfo->ReleaseLoc(testNode);
+    // Emit a zero trip test for the original outer-loop if the outer-loop
+	// has a condition
```
Comment (boingoing) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: Looks like these are tabs, you'll have to convert them to spaces to pass CI.

**Comment by boingoing** [comment]
Looks reasonable to me. Can you add a unit test to verify the fix?

**Comment by jackhorton** [comment]
Also, you may want to edit the PR description to "Fixes" rather than "GitHub issue" -- when you merge a PR with "Fixes #X" in the description (or in any of the commit messages, for that matter), it will auto-close the issue, which is useful.

File: test/loop/loopinversion.js [hunk 2]
```diff
@@ -70,5 +70,10 @@ function g5()
 };
 g5();
 
+// Tests that loop inversion does not crash when the outer loop does not have
+// a condition. In order to not stay in an infinite loop, these functions return
+// immediately. The bytecode will still be generated.
+function g6(p){if(!p) return; var a = 0; for (var h = 0;; ++h){ for (var r = 0; r < 1; ++r){ a = r; }}};g4a(0);
```
Comment (boingoing) [inline test/loop/loopinversion.js]: This line calls `g4a(0)`, is it meant to call `g6(0)` ?

Comment (wyrichte) [reply test/loop/loopinversion.js]: 95c41a5f0f9c4492673c08537b7d8d65d32c2942

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp (hunk 1 above)
Comment (wyrichte) [reply lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: 1335ae4ba0c4d195a3a3fba3e49420cba18c5276

**Pull request merged**